cpp/src/gandiva/function_registry_datetime.cc

```diff
@@ -0,0 +1,83 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gandiva/function_registry_datetime.h"
+#include <utility>
+#include <vector>
+
+namespace gandiva {
+
+void FunctionRegistryDateTime::GetDateTimeFnSignature(SignatureMap* map) {
+  // list of registered native functions.
+
+  static NativeFunction date_time_fn_registry_[] = {
+      // date/timestamp operations
+      DATE_TYPES(EXTRACT_SAFE_NULL_IF_NULL, extractMillennium),
```

Comment:
Make this a macro.

```C++
#define DATE_PART_FNS(name) \
  DATE_TYPES(EXTRACT_SAFE_NULL_IF_NULL, name##Millennium), \
  DATE_TYPES(EXTRACT_SAFE_NULL_IF_NULL, name##Century), \
  ...

DATE_PART_FNS(extract),
DATE_PART_FNS(date_trunc_),
```

Reply:
+1

Reply:
Incorporated the feedback in the latest commit.
Hope the change meets some of the expectations of the 'clean macros'. If the changes look okay my humble request is to approve the changes.

For further refactoring, we may pursue it separately on a case by case basis?